File: commands/cards/remove.ts
import { join } from "@std/path";
import prompts from "prompts";
import { getConfig } from "../../config.ts";
import { Deck } from "../../types.d.ts";

export const remove = async (deckSlug?: string, id?: string) => {
    const config = await getConfig();

    deckSlug = deckSlug ??
        ((
            await prompts({
                type: "autocomplete",
                name: "slug",
                message: "In which deck do you want to remove a card?",
                choices: Object.keys(config.decks).map((slug) => ({
                    title: slug,
                    value: slug,
                })),
            })
        ).slug as string);

    let deck: Deck;
    try {
        deck = JSON.parse(
            await Deno.readTextFile(
                join(config.decks[deckSlug].path, "deck.json"),
            ),
        );
    } catch (err) {
        if (err instanceof Deno.errors.NotFound) {
            console.log("Deck not found at path");
            return;
        }
        throw err;
    }

    if (!config.decks[deckSlug]) {
        console.log("Deck not found");
        return;
    }

    id = id ??
        ((
            await prompts({
                type: "autocomplete",
                name: "id",
                message: "Card ID",
                choices: Object.keys(deck.cards).map((id) => ({
                    title: id,
                    value: id,
                })),
            })
        ).id as string);

    if (!deck.cards[id]) {
        console.log("Card not found");
        return;
    }

    delete deck.cards[id];

    const continueWithRemove = (
        await prompts({
            type: "confirm",
            name: "continue",
            message: "Are you sure you want to remove this card?",
        })
    ).continue;

    if (!continueWithRemove) return;

    await Deno.writeTextFile(
        join(config.decks[deckSlug].path, "deck.json"),
        JSON.stringify(deck, null, 4),
    );
};

File: src/commands/cards/remove.rs
use std::{fs, path::Path};

use clap::Parser;
use inquire::{Confirm, Select, Text};

use crate::{config::get_config, deck::Deck};

#[derive(Parser, Debug, Clone)]
pub struct RemoveArgs {
    pub id: Option<String>,
}

pub fn run(_deck_id: Option<String>, args: RemoveArgs) {
    let config = get_config();

    let deck_id = match _deck_id {
        Some(id) => {
            if !config.decks.iter().any(|deck| deck.id == id) {
                panic!("Deck not found: {}", id);
            }

            id
        }
        None => match Select::new("Select a deck", config.decks.clone()).prompt() {
            Ok(selection) => selection.id,
            Err(err) => {
                panic!("Could not select deck: {}", err);
            }
        },
    };

    let deck_entry = config.decks.iter().find(|deck| deck.id == deck_id).unwrap();

    // Check if the deck is in path
    let deck_path = Path::new(deck_entry.path.as_str()).join("deck.json");
    if !deck_path.exists() {
        panic!("Deck not found in path: {}", deck_path.display());
    }

    let mut new_deck = match fs::read_to_string(&deck_path) {
        Ok(contents) => match serde_json::from_str::<Deck>(&contents) {
            Ok(deck) => deck,
            Err(err) => {
                panic!("Could not parse deck file: {}", err);
            }
        },
        Err(err) => {
            panic!("Could not read deck file: {}", err);
        }
    };

    let id = match args.id {
        Some(id) => id,
        None => match Text::new("Enter a unique ID for the card").prompt() {
            Ok(id) => id,
            Err(err) => {
                panic!("Could not get ID for card: {}", err);
            }
        },
    };

    new_deck.cards.retain(|card| card.id != id);

    if match Confirm::new("Are you sure you want to remove this card?").prompt() {
        Ok(confirmation) => confirmation,
        Err(err) => {
            panic!("Could not get confirmation: {}", err);
        }
    } {
        match fs::write(&deck_path, serde_json::to_string(&new_deck).unwrap()) {
            Ok(_) => {}
            Err(err) => {
                panic!("Could not write deck file: {}", err);
            }
        }
    }
}
